### ISSUE TITLE
OpenTK rewrites the DLL search order on Windows environments

### ISSUE THREAD

**Issue by glopesdev** [opentk/opentk#1443]
This is a regression introduced by #1223 and #1312.

It is unclear to me why the merged changes are necessary and I propose that they be reverted.

The motivation is to allow support for loading architecture-specific dependencies. While this is a common challenge which I have faced myself, there is already a simple way of specifying where exactly native DLLs are searched in Windows: https://docs.microsoft.com/en-us/windows/win32/dlls/dynamic-link-library-search-order

Essentially, any folder in the PATH environment variable will be scanned in order, so all an application needs to do is to check at runtime what the architecture is and modify it's own local PATH variable accordingly. The up-side is it will automatically work for all native dependencies and it can be easily made flexible to the particular situation of the application.

It's true that I have seen this pattern used in other C# packages, but I find it really bad practice. The problem is that different libraries have different "conventions" about where native DLLs should be scanned, and if each of them is modifying the global environment to make it work the way they think it should work, you will end up with native dependencies scattered all over the place, and worse you can end up with local environments which are impossible to resolve.

This is a breaking change for applications which have their own conventions about where DLLs are scanned, since OpenTK is effectively rewriting the global environment.

### Expected behavior

OpenTK should scan native DLLs using built-in platform-specific conventions.

### Actual behavior

In Windows, OpenTK is rewriting the native Dll search path by calling `SetDllDirectory`:
https://github.com/opentk/opentk/blob/803d8ba0d78369cbcd9b53fec559c4bc17c586c6/src/OpenTK/Toolkit.cs#L163

### Related information 

* Operating system: Windows
* What version of OpenTK: 3.3.2
* What runtime:  .NET Framework

**Comment by leezer3** [comment]
So, this is a very common practice in C# (as you note).

SetDllDirectory *only* applies to the calling process, and is not system-wide as you seem to think.
https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-setdlldirectorya
```
Adds a directory to the search path used to locate DLLs for the application.
```

It's vanishingly unlikely that someone will be attempting to load a different version of a component from a separately imported library (and AFAIK this won't work anyway)

Modifying the PATH variable is just another way to get the same effects.

**Comment by glopesdev** [comment]
I understand that it only applies to the calling process, the problem is that OpenTK is a library, and not an application, and so by using OpenTK 3.3.2 we are now effectively losing control over where exactly native DLLs are searched, and in what order.

Modifying the PATH or using `SetDllDirectory` is best left to the application layer and not to libraries.

**Comment by NogginBops** [comment]
I agree that if there are better way of specifying dll search directories we shouldn't be setting the dll search directory.

@glopesdev if you can make a PR that changes the documentation file to contain an updated guide without this feature, and remove the feature I can merge that.

**Comment by leezer3** [comment]
Changing the PATH variable will have *exactly* the same effects he's complaining about here.
The PATH variable is again common to the application.

As far as I can see, there's absolutely no way to actually get make what he wants work without shipping a deliberate x32 and x64 build in differing folders, each containing a copy of OpenTK & all other DLLs.

If you then start importing other libraries, the mess gets bigger and bigger, and makes a total mockery of the AnyCPU architecture in .Net

Sure, this solution isn't perfect, but it's the best you can do with the Win32 API

**Comment by NogginBops** [comment]
You can call `SetDLLDirectory` yourself can you not?
And if that isn't a nice solution, we should have a feature switch for this behavior so you can opt out.

**Comment by glopesdev** [comment]
> You can call SetDLLDirectory yourself can you not?

@NogginBops yup, this is exactly how we are using OpenTK and OpenAL right now, we configure the environment with our own PATH structure at application setup time. For us this is critical since our runtime includes a plugin system built on top of NuGet with its own set of conventions of where native dependencies are stored (e.g. using `runtime.json` for .NET Core or others).

A feature switch would be acceptable as a workaround, but in general it might lead to more confusion, especially if it's opt-out instead of opt-in. In the end, though, I really feel that these kinds of decisions are best left outside the scope of libraries. If OpenTK was a game engine or a runtime, such choices might be acceptable or even desirable, but given the stated aim of keeping the scope as a library, I think it would be more appropriate to leave this out.

@leezer3 can you elaborate on your specific use case so we can try to work out an elegant solution for everyone (e.g. are you developing a library or an application)? I am fully aware of the pains of dealing with architecture-specific dependencies for both libraries, runtimes and applications, and I've had to deal with several similar situations in the past.

**Comment by NogginBops** [comment]
If we where to introduce a feature switch it would be opt-in, disabled by default. It should only be a convenience thing where you don't have to setup such a system yourself.

**Comment by leezer3** [comment]
Good point, doing that yourself does work.
This however means that the developer has to deliberately do this, document it (as he can use whatever non-standard path convention he likes) and so-on.

With it done in the library side [and backwards compatible- this doesn't break anything], this means that the end-user can in theory drop in a new version of OpenTK & the appropriate DLL if they find it necessary. 

However, I'd stil argue it's a total non-issue. We're only adding a new path to be evaluated first in the current process search, and if this fails, it still falls down the standard Win32 search tree.

TBQH, I'd place OpenTK far closer to an engine than a library. Gets rather into semantics :/


@glopesdev 
I'm the maintainer / primary developer for https://openbve-project.net/
Various issues, primarily related to OpenAL / SDL:
* System OpenAL is often spotty or missing.
* Later versions of OpenAL Soft are broken with OpenTK. (I haven't got the specific version numbers to hand)
* Creative OpenAL has issues over a certain number of concurrent sounds (Again, haven't got specifics to hand, but at one stage we distributed thier redistributable, so many longer term users may well have this on the system)
* x86 and x64 versions are absolutely required due to legacy C++ plugins. Whilst there is actually a NamedPipes proxy app to allow x86 plugins to be used on x64, many users prefer the x32 native version. We've also got some x86 only .Net plugins floating around in the mix for good measure, which require the x86 build :/

Sure, I can absolutely set the DLL directory path myself, and I'm not specificallly averse to doing that.
Without doing this though, we get straight back to the mess.


My point is somewhat more that this is practically an unofficial convention, used by many other libraries, not reinventing an entirely new and undocumented wheel.
Ultimately, I suppose the maintainers (not me!) need to decide whether ease of use / unofficial convention works better for the silent majority, most of whom I would suspect are barely past `HelloWorld()` not running a plugin system on top of nuget!

**Comment by leezer3** [comment]
See, there's another problem (rod for our own backs).....

Our target .Net version is 4.0 to run [everything bar the 'new' train editor] / 4.7 to build + new train editor (rather more technically, the run target is mono 5.10 [**freebsd**], build target is mono 5.18 [**debian oldstable**])
Newer .Net versions have all sorts of new and shiny stuff, but unfortunately that takes a massive amount of time to filter down into the more conservative distros.

I haven't got the .Net target version for OpenTK 3x off the top of my head, but it's certainly 4.0 or lower. I'd be totally against increasing this just for something like this.

---

I'd completely agree that there's no real Microsoft standard, but an x86 / x64 folder beneath your DLL is as close to an informal standard as we've got.
I'm still very much of the opinion that DLL-hell fixing should be done in OpenTK, if nothing else than because the user is not touching these files directly. If the end-user specifically was doing a DLLImport on **openal32.dll** , then finding the appropriate version & fixing DLL-hell should be the user's problem.
In this case, however, OpenTK is doing whatever it does with the DLL transparently to the user, hence IMHO it's the frameworks job to locate it appropriately.

**Comment by NogginBops** [comment]
I think we should add a feature swtich somewhere where you can turn on this behavior, and having it off by default.
We should document the feature well so people can find it though.

**Comment by Barbosik** [comment]
I think this is very bad idea to insert DLL search path into library. Such a code should be made in the main app executable, but not in the library. I'm using my custom own (overloaded) DLL search logic, so if such a code will be placed into library it will lead to errors due to conflicted DLL search path logic between main app executable and OpenTK library.

**Comment by NogginBops** [comment]
@glopesdev @Barbosik @leezer3  does #1473 look like a good solution for this?

**Comment by NogginBops** [comment]
Fixed in #1473
